### PR TITLE
change file merging algorithm to take into account open file descriptors limitation

### PR DESCRIPTION
**Pull request by sszuev**
#9

### CODE REVIEW

File: src/main/kotlin/files/FileMerge.kt [hunk 1]
```diff
@@ -83,6 +86,68 @@ fun mergeFilesInverse(
...
+    target: Path,
+    comparator: Comparator<ByteArray> = byteArrayStringComparator().reversed(),
+    delimiter: ByteArray = "\n".toByteArray(Charsets.UTF_8),
+    bomSymbols: ByteArray = byteArrayOf(),
+    allocatedMemorySizeInBytes: Int = 2 * MERGE_FILES_MIN_WRITE_BUFFER_SIZE_IN_BYTES,
+    writeToTotalMemRatio: Double = MERGE_FILES_WRITE_BUFFER_TO_TOTAL_MEMORY_ALLOCATION_RATIO,
+    coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.IO) + CoroutineName("mergeFilesInverse")
+) {
+    require(sources.size > 1) { "Number of given sources (${sources.size}) must greater than 1" }
+    require(writeToTotalMemRatio > 0.0 && writeToTotalMemRatio < 1.0)
```
Comment (githubuser-byte) [inline src/main/kotlin/files/FileMerge.kt]: an exception message can also be added here, "writeToTotalMemRatio must fall within the range of 0 to 1"

File: src/main/kotlin/files/MergeSort.kt [hunk 2]
```diff
@@ -322,6 +347,169 @@ suspend fun suspendSort(
...
+    //      val chunk = reversed.removeChunk()
+    //      val res = mergeFilesInverse(chunk)
+    //      direct.add(res)
+    //  }
+    //  if (direct.size > 1) {
+    //      val chunk = direct.removeChunk()
+    //      val res = mergeFilesInverse(chunk)
+    //      reverse.add(res)
+    //  }
+    // }
```
Comment (githubuser-byte) [inline src/main/kotlin/files/MergeSort.kt:433]: commented code

Comment (sszuev) [reply src/main/kotlin/files/MergeSort.kt:433]: This is pseudocode. It was supposed to help understanding the algorithm (which is actually far from perfect).

**Comment by sszuev** [comment]
merged by cherrypicks

**Pull request closed without merging**